node/runtime/src/lib.rs

```diff
@@ -985,136 +988,126 @@ impl Runtime {
-        let mut new_receipts = vec![];
-        let mut state_update = StateDbUpdate::new(self.state_db.clone(), apply_state.root);
-        let mut authority_proposals = vec![];
-        let shard_id = apply_state.shard_id;
-        let block_index = apply_state.block_index;
-        let mut tx_result = vec![];
-        for tx in prev_receipts.iter().chain(transactions) {
-            tx_result.push(Self::process_transaction(
-                self,
-                &mut state_update,
-                shard_id,
+    fn process_receipt(
+        runtime: &mut Self,
+        state_update: &mut StateDbUpdate,
+        shard_id: ShardId,
+        block_index: BlockIndex,
+        receipt: &ReceiptTransaction,
+        new_receipts: &mut HashMap<ShardId, Vec<ReceiptTransaction>>,
+    ) -> TransactionResult {
+        let mut result = TransactionResult::default();
+        if account_to_shard_id(&receipt.receiver) == shard_id {
+            let mut tmp_new_receipts = vec![];
+            let apply_result = runtime.apply_receipt(
+                state_update, 
+                receipt,
+                &mut tmp_new_receipts,
                 block_index,
-                tx,
-                &mut new_receipts,
-                &mut authority_proposals
-            ));
+                &mut result.logs
+            );
+            for receipt in tmp_new_receipts {
+                result.receipts.push(receipt.nonce);
+                let shard_id = receipt.shard_id();
+                if new_receipts.contains_key(&shard_id) {
+                    new_receipts
+                    .entry(shard_id)
+                    .and_modify(|e| e.push(receipt));
+                } else {
+                    new_receipts.insert(shard_id, vec![receipt]);
+                }
+            }
+            match apply_result {
+                Ok(()) => {
+                    state_update.commit();
+                    result.status = TransactionStatus::Completed;
+                }
+                Err(s) => {
+                    debug!(target: "runtime", "{}", s);
+                    state_update.rollback();
+                    result.status = TransactionStatus::Failed;
+                }
+            };
+            result
+        } else {
+            // wrong receipt
+            debug!(target: "runtime", "receipt sent to the wrong shard");
+            result.status = TransactionStatus::Failed;
+            result
         }
-        let (db_transaction, new_root) = state_update.finalize();
-        Some((db_transaction, new_root, tx_result))
     }
 
-    /// apply receipts from previous block and transactions and receipts from this block
+    /// apply receipts from previous block and transactions from this block
     pub fn apply(
         &mut self,
         apply_state: &ApplyState,
-        prev_receipts: &[Transaction],
-        transactions: &[Transaction],
+        prev_receipts: &[ReceiptTransaction],
```

Comment:
why not [ReceiptBlock]?

Reply:
Could not avoid circular dependency if I use `ReceiptBlock` here. Might consider using generic type parameter to bypass that

Reply:
Fixed in 2e6dd8c